# MNIST, Le Hello World du deep learning !

In [1]:
import tensorflow

### Import du jeu de données "mnist" intégré directement à tensorflow, puis chargement des données dans les variables
Le jeu d'entrainement MNIST est composé d'un jeu d'entrainement de 60 000 entrées labelisées, pour savoir quels sont les résultats attendus et que l'algorithme s'entraine avec ainsi que d'un jeu de test composé de 10 000 entrées labelisées aussi, avec lequel je vais pouvoir évaluer la précision du modèle en comparant sa réponse au label fourni

Chaque entrée est un tableau de nombres ayant une valeur entre 0 et 1 représentant les pixels de l'image. Chaque image réprésente un chiffre entre 0 et 9 que notre modèle va s'entrainer à reconnaitre avec le jeu d'entrainement, puis que je vais évaluer avec le jeu de test

In [2]:
from tensorflow.keras.datasets import mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

In [3]:
train_images.shape

(60000, 28, 28)

In [4]:
len(train_labels)

60000

In [5]:
test_images.shape

(10000, 28, 28)

In [6]:
len(test_labels)

10000

### Import de l'architecture

Nous allons maintenant importé les couches ou "layers" pour filtrer nos données. Des données rentre, puis ressortent dans une forme qui nous est plus utile. La plupart des algorithme de deep learning consistent à enchainer plusieurs layers simples qui vont implémenter une forme de "distillation des données" tel un tamis.

In [7]:
from tensorflow import keras
from tensorflow.keras import layers
model = keras.Sequential([
layers.Dense(512, activation="relu"),
layers.Dense(10, activation="softmax")
])

Ici nous avons 2 "Dense" layers connectée. La 2e layers est une classification softmax à 10 scores. Elles va donc renvoyer 10 scores de probabilité (qui s'additionnent jusqu'à un total de 1) chaque score étant la probabilité que le chiffre de l'image appartienne à une de ces 10 valeurs.

In [8]:
model.compile(optimizer="rmsprop",
loss="sparse_categorical_crossentropy",
metrics=["accuracy"])

Maintenant préparer notre modèle à l'entrainement il nous faut 3 choses supplémentaires:
 - Un optimiseur, qui va définir comment le modèle va s'adapter aux données qu'il vois pour améliorer sa performance
 - une fonction de perte (loss), ou comment le modèle va pouvoir évaluer sa performance et ainsi partir dans la bonne direction
 - et des valeurs pour surveiller pendant l'entrainement et le test. ici on va simplement se contenter de la précision (accuracy)

In [9]:
train_images = train_images.reshape((60000, 28 * 28))
train_images = train_images.astype("float32") / 255
test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype("float32") / 255

Avant l'entrainement je vais modifier la forme des données pour qu'elles soient lisibles par notre modèle. Je vais d'abord changer le typage des données, passant de uint8 à float32 puis je passe de valeurs entre 0 et 255, à des valeurs entre 0 et 1. 

Je suis maintenant prêt à commentcer l'entrainement du modèle. Avec Keras cela se fait avec un appel à la fonction "fit()" qui va "fit" mon modèle aux données d'entrainement

In [10]:
model.fit(train_images, train_labels, epochs=5, batch_size=128)

Epoch 1/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8776 - loss: 0.4322
Epoch 2/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9658 - loss: 0.1155
Epoch 3/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9797 - loss: 0.0707
Epoch 4/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9845 - loss: 0.0510
Epoch 5/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9893 - loss: 0.0375


2 quantités sont affichées pendant l'entrainement, la perte du modèle sur les données d'entrainement, et la précision du modèle sur les données d'entrainement.
Maintenant que le modèle est entrainé, je peux l'utiliser pour "prédire" la valeur d'un chiffre à partir de données qui ne sont pas dans le jeu d'entrainement.

In [11]:
test_digits = test_images[0:20]
predictions = model.predict(test_digits)
predictions[0]

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 43ms/step


array([1.4896453e-07, 8.8369267e-09, 1.0527246e-05, 1.3607496e-04,
       9.6368719e-11, 1.1626091e-07, 2.2105848e-11, 9.9984753e-01,
       3.4124312e-07, 5.1913394e-06], dtype=float32)

Chaque nombre dans ce tableau prediction[0] représente la probabilité que chaque chiffre corresponde à l'image. Ici la probabilité la plus haute est à l'index numéro 7, selon notre modèle cela doit donc être un 7

In [12]:
predictions[0].argmax()

7

In [13]:
predictions[0][7]
0.99999106

0.99999106

In [14]:
test_labels[0]

7

On peut vérifier ici dans le label si c'est bien un 7

In [15]:
test_loss, test_acc = model.evaluate(test_images, test_labels)

313/313 ━━━━━━━━━━━━━━━━━━━━ 0s 745us/step - accuracy: 0.9783 - loss: 0.0697


J'appelle la fonction evaluate() pour tester mon modèle sur les données de test et je récupère les valeurs de perte et de précision

In [16]:
print(f"test_acc: {test_acc}")

test_acc: 0.9822999835014343


La précision du jeu de test est de 97.8%, plus bas qu'avec notre jeu d'entrainement à 98.9%